### ISSUE TITLE
Allow Git blame to be customizable

### ISSUE THREAD

**Issue by hshreekar** [zed-industries/zed#10771]
Summary: Support custom definition of what commit data shows in git blame gutter

### Description the feature

The default git blame takes up too much space (in my opinion). 

Allow the user to change the behaviour of git blame, allowing users to choose what entry he wishes to see (The full information may still be seen by hovering and clicking on it ). This may also allow users to configure the author width as seen in issue #10631. 

Run: `git blame --porcelain FILENAME` to the example fields which might be supported:
- author / author-mail / author-time / author-tz
- committer / committer-mail / author-time / author-tz
- summary
- previous commit hash

Assumably it would be nice to have some date formatting/conversion options for author-time / author-tz.

```json
{
  "git": {
    "gutter_format": "{shorthash} {author} {relative(author-time)}"
  }
}
```
For something like the current format:

<img width="364" alt="Image" src="https://github.com/user-attachments/assets/9bd13364-7f88-4525-9eed-2611436892c4" />

**Comment by hshreekar** [comment]
I am willing to submit a PR for this, incase this makes sense and someone who could guide me 

fields i am thinking would make sense are , 
```
show_author: bool, 
show_relative_timestamp: bool,
show_commit_id: bool,
author_style: an enum containing ("dynamic", "fixed_width") (maybe with a better name ),
color_author_by_commit: bool  (add a  filled circle after the author name coloured by commit id), 
dock: enum of right or left
```

**Comment by mrnugget** [comment]
Do you mean the inline git blame or the git blame in the gutter?

**Comment by hshreekar** [comment]
Git Blame in the Gutter

**Comment by JosephTLyons** [comment]
One thing that might help with the width of the git blame gutter is dynamically computing what is needed:

- https://github.com/zed-industries/zed/issues/10631#issuecomment-2059553516

**Comment by hshreekar** [comment]
Yes that would help with the width of the git blame , I only skimmed through the git blame code portion but I believe we can add a field for max_author_len in the blame struct or have a field called authors which would be  vec<string> and compute the max length just before rendering the elements. I would prefer the authors field just because we could maybe reuse the blame entries to implement specialized file searches ( something like search for files edited by a particular author) although maybe it's too compute intensive 

I still believe we can maybe make the info in git blame gutter more dense like not showing commit id or just replacing the string with a custom colour filled circle (my preference) or colour the line number by the commit id (might look ugly, also some users might not have enabled line numbers ) 

However I do believe that if most of the users of zed are fine with the current method, only the author width can be adjusted.

**Comment by jansol** [comment]
One nice option would be to put commit, author and date on separate lines when possible and have a colored bar on one side to indicate which lines are part of the same blame chunk.

The GitHub blame view actually does this last bit so should be a good reference.

**Comment by mrnugget** [comment]
> One nice option would be to put commit, author and date on separate lines when possible and have a colored bar on one side to indicate which lines are part of the same blame chunk.

That's actually not that simple, since right now we "simply" draw next to the editor line. If you have multiple blame-lines you also need to space out the editor lines, which (I think) is not as easy as making the format of the blame lines configurable.

**Comment by feketegy** [comment]
Would it be possible to change (or override the theme settings) of the foreground color of the inline git blame?

I'm currently using the Andromeda theme and the inline blame color is pretty distracting.

![Pasted_Image_8_20_24__12_35](https://github.com/user-attachments/assets/a1127db8-181c-4f35-8162-6f995f73fed2)

**Comment by mrnugget** [comment]
@feketegy not separately, yet, but you can define a color like this:

```json
{
  "experimental.theme_overrides": {
    "hint": "#CCCCCC"
  }
}
```

**Comment by alexpyattaev** [comment]
I would argue the most important thing would be to support -M and -C flags of git blame, such that when code is moved around the correct author is still shown.

**Comment by olejorgenb** [comment]
> I would argue the most important thing would be to support -M and -C flags of git blame, such that when code is moved around the correct author is still shown.

Draft PR: #44333

**Comment by olejorgenb** [comment]
In case someone find this issue through search: You can use the [git mailmap](https://git-scm.com/docs/gitmailmap) feature to shorten author names independent of Zed.

<img width="474" height="117" alt="Image" src="https://github.com/user-attachments/assets/0cf01d88-a069-4cec-9e5d-4f7e4a4a6e5b" />
